<a href="https://colab.research.google.com/github/Obanai4you/CSIT-LAB-WORKS/blob/main/Dataware_Housing_And_Datamining/Lab2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#LAB 2:
******
#Define Data Quality:

Data Quality refers to how accurate, reliable, and useful a set of data is for its intended purpose.

In simple terms, high-quality data means the data is correct, complete, consistent, and up-to-date, so it can be trusted for decision-making.

*****

#Issue with Data Quality: (In Data Mining & Data Warehousing):

1. Incomplete Data – Missing values or fields
2. Noisy Data – Errors, outliers, or random variations
3. Inconsistent Data – Conflicting information across sources
4. Duplicate Data – Same records repeated multiple times
5. Outdated Data – Data not updated regularly
6. Data Integration Issues – Mismatch from multiple sources
7. Data Entry Errors – Mistakes during data input

*****

#Missing Values and Its Reasons:

1. Data not collected – Some fields are left blank during data collection
2. Human errors – Mistakes while entering data
3. Equipment/system failure – Sensors or systems fail to record data
4. Data corruption – Data lost due to technical issues
5. Non-response – Users choose not to provide certain information.

******

#Techniques to Deal with Missing Values:
1. Ignore/Delete records – Remove rows with missing values (if few)
2. Manual filling – Fill missing values by hand (not practical for large data)
3. Use global constant – Replace with a value like “Unknown” or 0
4. Mean/Median/Mode substitution – Fill using average or most frequent value
5. Use attribute mean (class-based) – Mean calculated within same category/class
6. Prediction methods – Use algorithms (like regression, decision trees) to estimate


##1. Dataset Understanding
******
The dataset contains 1025 rows and 14 columns.
It is related to heart disease prediction.
The target variable (target) indicates:
0 → No disease
1 → Disease present

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

In [2]:
heart_ds = pd.read_csv("/content/heart.csv")

In [3]:
heart_ds.shape

(1025, 14)

In [5]:
heart_ds.head()
heart_ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB


In [6]:
heart_ds.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0
1,53,1,0,140,203,1,0,155,1,3.1,0,0,3,0
2,70,1,0,145,174,0,1,125,1,2.6,0,0,3,0
3,61,1,0,148,203,0,1,161,0,0.0,2,1,3,0
4,62,0,0,138,294,1,1,106,0,1.9,1,3,2,0


##2. Checking Missing Values
**********
Missing values reduce model accuracy.
They can occur due to:

a. Data entry errors

b. Equipment failure

c. Human mistakes

In [7]:
heart_ds.isnull().sum()

,0
age,0
sex,0
cp,0
trestbps,0
chol,0
fbs,0
restecg,0
thalach,0
exang,0
oldpeak,0


##Interpretation
************
a. The output of heart_ds.isnull().sum() shows that all columns have 0 missing values

b. This means there are no null or missing entries in the dataset.

###Final Statement
********
a. Since no missing values are present, no data cleaning is required for missing values.

b. The dataset is complete and reliable for further analysis.

##3. Handling Missing Values
Techniques:

a. Remove rows → dropna()

b. Fill with mean/median

c. Fill with mode (categorical)

##4. Checking Duplicate Values

Duplicate values are rows that appear more than once in the dataset.
They can:

a.Bias the analysis

b.Reduce model performance

Hence, they must be identified and removed.

In [10]:
heart_ds.duplicated()

,0
0,False
1,False
2,False
3,False
4,False
...,...
1020,True
1021,True
1022,True
1023,True


In [11]:
heart_ds.duplicated().sum()

np.int64(723)

###Interpretation
***********
a.The output shows that there are 723 duplicate rows present in the dataset.

b.This is a significant number of duplicates and can negatively impact data analysis and model performance.

###Final Statement
************

a.Duplicate rows can lead to biased results and incorrect insights.

b.Therefore, it is necessary to remove these duplicate entries to ensure data quality.


##5. Eliminating Dubllicated Rows:
***************
Removing duplicate values ensures that each observation in the dataset is unique.
This improves the reliability and accuracy of the analysis.

In [12]:
heart_ds = heart_ds.drop_duplicates().reset_index(drop=True)

In [14]:
heart_ds.duplicated()

,0
0,False
1,False
2,False
3,False
4,False
...,...
297,False
298,False
299,False
300,False


In [13]:
heart_ds.duplicated().sum()

np.int64(0)

After removing duplicates, the dataset now contains 0 duplicate rows.
The dataset is now clean and consistent.

##6. Data Types Checking
**************
a. Data types define the kind of values stored in each column.

b. Correct data types are important for proper analysis and model building.

c. Incorrect data types may lead to errors or inaccurate results.

In [15]:
heart_ds.dtypes

,0
age,int64
sex,int64
cp,int64
trestbps,int64
chol,int64
fbs,int64
restecg,int64
thalach,int64
exang,int64
oldpeak,float64


All columns are in appropriate numeric data types.
No data type conversion is required.
The dataset is properly formatted for analysis.

In [16]:
heart_ds.describe()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
count,302.00000,302.000000,302.000000,302.000000,302.000000,302.000000,302.000000,302.000000,302.000000,302.000000,302.000000,302.000000,302.000000,302.000000
mean,54.42053,0.682119,0.963576,131.602649,246.500000,0.149007,0.526490,149.569536,0.327815,1.043046,1.397351,0.718543,2.314570,0.543046
std,9.04797,0.466426,1.032044,17.563394,51.753489,0.356686,0.526027,22.903527,0.470196,1.161452,0.616274,1.006748,0.613026,0.498970
min,29.00000,0.000000,0.000000,94.000000,126.000000,0.000000,0.000000,71.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,48.00000,0.000000,0.000000,120.000000,211.000000,0.000000,0.000000,133.250000,0.000000,0.000000,1.000000,0.000000,2.000000,0.000000
50%,55.50000,1.000000,1.000000,130.000000,240.500000,0.000000,1.000000,152.500000,0.000000,0.800000,1.000000,0.000000,2.000000,1.000000
75%,61.00000,1.000000,2.000000,140.000000,274.750000,0.000000,1.000000,166.000000,1.000000,1.600000,2.000000,1.000000,3.000000,1.000000
max,77.00000,1.000000,3.000000,200.000000,564.000000,1.000000,2.000000,202.000000,1.000000,6.200000,2.000000,4.000000,3.000000,1.000000


#Conclusion:
The statistical summary shows that the dataset has a reasonable distribution of values.
No extreme anomalies are observed in the dataset.
The features appear suitable for further analysis.

The dataset was examined for data quality issues.
Missing values were checked and none were found.
A large number of duplicate values (723) were identified and successfully removed.
Data types were verified and found to be appropriate.
Statistical summary was analyzed to understand data distribution.
The dataset is now clean, consistent, and ready for further analysis or model building.